core/src/main/scala/kafka/utils/ZkUtils.scala

```diff
@@ -456,6 +457,58 @@ class ZkUtils(val zkClient: ZkClient,
   }
 
   /**
+    * Callback that processes result of asynchronous version of conditionalUpdatePersistentPathAsync
+    */
+  private class conditionalUpdatePersistentPathCallback extends StatCallback {
+
+    def processResult(rc: Int, path: String, internalCtx: scala.Any, stat: Stat): Unit = {
+      val (zkUtils, zkHandle, path2, data, expectVersion, optionalCheckerAsync, callback, ctx) =
+        internalCtx.asInstanceOf[(ZkUtils, ZooKeeper, String, String, Int,
+          Option[(ZkUtils, String, String, (Boolean, Int, scala.Any) => Unit, scala.Any) => Unit],
+          (Boolean, Int, scala.Any) => Unit, scala.Any)]
+
+      Code.get(rc) match {
+        case Code.OK =>
+          debug("Conditional (async) update of path %s with value %s and expected version %d succeeded, returning the new version: %d"
+            .format(path, data, expectVersion, stat.getVersion))
+          return callback(true, stat.getVersion, ctx)
+        case Code.BADVERSION =>
+          optionalCheckerAsync match {
+            case Some(checker) =>
+              return checker(zkUtils, path, data, callback, ctx)
+            case _ => debug("Checker method is not passed skipping zkData match")
+          }
+          warn("Conditional update (async) of path %s with data %s and expected version %d failed due to bad version".format(path, data,
+            expectVersion))
+          return callback(false, -1, ctx)
+        case Code.CONNECTIONLOSS | Code.SESSIONEXPIRED =>
+          // retry
+          warn("Conditional update (async) of path %s with data %s and expected version %d failed due to connection/session loss %d. Retrying.".format(path, data,
+            expectVersion, Code.get(rc).intValue()))
+          zkHandle.setData(path, ZKStringSerializer.serialize(data), expectVersion, new conditionalUpdatePersistentPathCallback, internalCtx)
```

Comment:
@fpj : I'm following the same invariant as the the synchronous version of conditionalUpdatePersistentPathAsync, called conditionalUpdatePersistentPath. The invariant is stated as "When there is a ConnectionLossException during the conditional update, zkClient will retry the update and may fail since the previous update may have succeeded (but the stored zkVersion no longer matches the expected one). In this case, we will run the optionalChecker to further check if the previous write did indeed succeeded." The synchronous version wraps the retry in writeDataReturnStat. This for both CONNECTIONLOSS and SESSIONEXPIRED.